File: ios/Classes/BannerAd/BannerAdLoader.swift
import Foundation
import GoogleMobileAds
import Flutter

/// Protocol for banner ad loader callbacks.
protocol BannerAdLoaderDelegate: AnyObject {
    func adLoader(_ loader: BannerAdLoader, didReceiveBannerAd bannerView: BannerView)
    func adLoader(_ loader: BannerAdLoader, didFailWithError error: Error)
}

/// Handles loading banner ads from AdMob.
class BannerAdLoader: NSObject {

    private let adUnitId: String
    let controllerId: String
    private let channel: FlutterMethodChannel
    private let enableDebugLogs: Bool
    private let adSize: AdSize

    private var bannerView: BannerView?

    weak var delegate: BannerAdLoaderDelegate?
    private var onAdLoadedCallback: ((BannerView) -> Void)?

    init(
        adUnitId: String,
        controllerId: String,
        channel: FlutterMethodChannel,
        adSize: AdSize,
        enableDebugLogs: Bool = false
    ) {
        self.adUnitId = adUnitId
        self.controllerId = controllerId
        self.channel = channel
        self.adSize = adSize
        self.enableDebugLogs = enableDebugLogs
        super.init()
    }

    /// Sets the callback for when an ad is loaded.
    func setOnAdLoadedCallback(callback: @escaping (BannerView) -> Void) {
        onAdLoadedCallback = callback
    }

    /// Loads a banner ad.
    func loadAd() {
        guard let rootVC = rootViewController() else {
            log("ERROR: Could not get root view controller!")
            return
        }

        let bannerView = BannerView(adSize: adSize)
        bannerView.adUnitID = adUnitId
        bannerView.rootViewController = rootVC
        bannerView.delegate = self

        self.bannerView = bannerView
        bannerView.load(Request())
    }

    /// Gets the currently loaded banner view.
    func getBannerView() -> BannerView? {
        return bannerView
    }

    /// Destroys the loader and releases resources.
    func destroy() {
        bannerView?.delegate = nil
        bannerView = nil
        onAdLoadedCallback = nil
    }

    // MARK: - Private Methods

    private func rootViewController() -> UIViewController? {
        return UIApplication.shared.connectedScenes
            .compactMap { ($0 as? UIWindowScene)?.windows.first }
            .first?.rootViewController
    }

    private func sendEvent(_ method: String, arguments: [String: Any]) {
        DispatchQueue.main.async { [weak self] in
            self?.channel.invokeMethod(method, arguments: arguments)
        }
    }

    private func log(_ message: String) {
        if enableDebugLogs {
            print("[BannerAdLoader][\(controllerId)] \(message)")
        }
    }
}

// MARK: - BannerViewDelegate

extension BannerAdLoader: BannerViewDelegate {

    func bannerViewDidReceiveAd(_ bannerView: BannerView) {
        sendEvent("onAdLoaded", arguments: [
            "controllerId": controllerId
        ])

        onAdLoadedCallback?(bannerView)
        delegate?.adLoader(self, didReceiveBannerAd: bannerView)
    }

    func bannerView(_ bannerView: BannerView, didFailToReceiveAdWithError error: Error) {
        log("Banner ad failed to load: \(error.localizedDescription)")

        let nsError = error as NSError
        sendEvent("onAdFailedToLoad", arguments: [
            "controllerId": controllerId,
            "error": error.localizedDescription,
            "errorCode": nsError.code
        ])

        delegate?.adLoader(self, didFailWithError: error)
    }

    func bannerViewDidRecordClick(_ bannerView: BannerView) {
        sendEvent("onAdClicked", arguments: [
            "controllerId": controllerId
        ])
    }

    func bannerViewDidRecordImpression(_ bannerView: BannerView) {
        sendEvent("onAdImpression", arguments: [
            "controllerId": controllerId
        ])
    }

    func bannerViewWillPresentScreen(_ bannerView: BannerView) {
        sendEvent("onAdOpened", arguments: [
            "controllerId": controllerId
        ])
    }

    func bannerViewWillDismissScreen(_ bannerView: BannerView) {
    }

    func bannerViewDidDismissScreen(_ bannerView: BannerView) {
        sendEvent("onAdClosed", arguments: [
            "controllerId": controllerId
        ])
    }

    func bannerView(_ bannerView: BannerView, didReceiveAdValue value: AdValue) {
        let valueInDollars = Double(truncating: value.value) / 1_000_000.0
        sendEvent("onAdPaid", arguments: [
            "controllerId": controllerId,
            "value": valueInDollars,
            "currencyCode": value.currencyCode
        ])
    }
}

File: android/src/main/kotlin/com/tqc/ads/flutter_admob_native_ads/banner/BannerAdLoader.kt
package com.tqc.ads.flutter_admob_native_ads.banner

import android.content.Context
import android.os.Handler
import android.os.Looper
import android.util.Log
import com.google.android.gms.ads.AdListener
import com.google.android.gms.ads.AdRequest
import com.google.android.gms.ads.AdSize
import com.google.android.gms.ads.AdView
import com.google.android.gms.ads.LoadAdError
import com.google.android.gms.ads.OnPaidEventListener
import io.flutter.plugin.common.MethodChannel

/**
 * Handles loading banner ads from AdMob.
 */
class BannerAdLoader(
    private val context: Context,
    private val adUnitId: String,
    private val controllerId: String,
    private val channel: MethodChannel,
    private val adSize: AdSize,
    private val enableDebugLogs: Boolean = false
) {
    companion object {
        private const val TAG = "BannerAdLoader"
    }

    private var adView: AdView? = null
    private var onAdLoadedCallback: ((AdView) -> Unit)? = null
    private val mainHandler = Handler(Looper.getMainLooper())

    /**
     * Sets the callback for when an ad is loaded.
     */
    fun setOnAdLoadedCallback(callback: (AdView) -> Unit) {
        onAdLoadedCallback = callback
    }

    /**
     * Loads a banner ad.
     */
    fun loadAd() {
        adView = AdView(context)
        adView?.adUnitId = adUnitId
        adView?.setAdSize(adSize)

        adView?.setOnPaidEventListener(OnPaidEventListener { adValue ->
            val valueInCurrency = adValue.valueMicros / 1_000_000.0
            sendEvent("onAdPaid", mapOf(
                "controllerId" to controllerId,
                "value" to valueInCurrency,
                "currencyCode" to adValue.currencyCode
            ))
        })

        adView?.adListener = object : AdListener() {
            override fun onAdLoaded() {
                adView?.let { onAdLoadedCallback?.invoke(it) }
                sendEvent("onAdLoaded", mapOf("controllerId" to controllerId))
            }

            override fun onAdFailedToLoad(error: LoadAdError) {
                log("Banner ad failed to load: ${error.message} (code: ${error.code})")
                sendEvent("onAdFailedToLoad", mapOf(
                    "controllerId" to controllerId,
                    "error" to (error.message ?: "Unknown error"),
                    "errorCode" to error.code
                ))
            }

            override fun onAdClicked() {
                sendEvent("onAdClicked", mapOf("controllerId" to controllerId))
            }

            override fun onAdImpression() {
                sendEvent("onAdImpression", mapOf("controllerId" to controllerId))
            }

            override fun onAdOpened() {
                sendEvent("onAdOpened", mapOf("controllerId" to controllerId))
            }

            override fun onAdClosed() {
                sendEvent("onAdClosed", mapOf("controllerId" to controllerId))
            }
        }

        adView?.loadAd(AdRequest.Builder().build())
    }

    /**
     * Gets the currently loaded banner view.
     */
    fun getAdView(): AdView? = adView

    /**
     * Destroys the loader and releases resources.
     */
    fun destroy() {
        adView?.destroy()
        adView = null
        onAdLoadedCallback = null
    }

    /**
     * Sends an event to Flutter via method channel.
     */
    private fun sendEvent(method: String, arguments: Map<String, Any?>) {
        val emit = {
            try {
                channel.invokeMethod(method, arguments)
            } catch (e: Exception) {
                log("Error sending event $method: ${e.message}")
            }
        }
        if (Looper.myLooper() == Looper.getMainLooper()) {
            emit()
        } else {
            mainHandler.post { emit() }
        }
    }

    private fun log(message: String) {
        if (enableDebugLogs) {
            Log.d(TAG, "[$controllerId] $message")
        }
    }
}
